src/main/java/org/dependencytrack/tasks/BomUploadProcessingTask.java

```diff
@@ -84,174 +114,715 @@ public BomUploadProcessingTask() {
+                        // FIXME: Add reference to BOM after we have dedicated BOM server
+                        .subject(new BomProcessingFailed(ctx.project, /* bom */ "(Omitted)", ex.getMessage(), ex.ctx.bomFormat, ex.ctx.bomSpecVersion)));
+            } catch (Exception ex) {
+                LOGGER.error("BOM processing failed unexpectedly (%s)".formatted(ctx), ex);
+                kafkaEventDispatcher.dispatchAsync(ctx.project.getUuid(), new Notification()
+                        .scope(NotificationScope.PORTFOLIO)
+                        .group(NotificationGroup.BOM_PROCESSING_FAILED)
                         .level(NotificationLevel.ERROR)
+                        .title(NotificationConstants.Title.BOM_PROCESSING_FAILED)
                         .content("An error occurred while processing a BOM")
                         // FIXME: Add reference to BOM after we have dedicated BOM server
-                        .subject(new BomProcessingFailed(bomProcessingFailedProject, /* bom */ "(Omitted)", ex.getMessage(), bomProcessingFailedBomFormat, bomProcessingFailedBomVersion)));
+                        .subject(new BomProcessingFailed(ctx.project, /* bom */ "(Omitted)", ex.getMessage(), ctx.bomFormat /* (may be null) */, ctx.bomSpecVersion /* (may be null) */)));
             } finally {
                 timerSample.stop(TIMER);
-                qm.commitSearchIndex(true, Component.class);
-                qm.commitSearchIndex(true, ServiceComponent.class);
-                qm.close();
             }
         }
     }
 
-    private void processComponent(final QueryManager qm, Component component,
-                                  final List<Component> flattenedComponents,
-                                  final List<Component> newComponents) {
-        final boolean isNew = component.getUuid() == null;
-        component.setInternal(InternalComponentIdentificationUtil.isInternalComponent(component, qm));
-        component = qm.createComponent(component, false);
-        final long oid = component.getId();
-        // Refreshing the object by querying for it again is preventative
-        component = qm.getObjectById(Component.class, oid);
-        flattenedComponents.add(component);
-        if (isNew) {
-            newComponents.add(qm.detach(Component.class, component.getId()));
+    private void processBom(final Context ctx, final File bomFile) throws BomProcessingException {
+        LOGGER.info("Consuming uploaded BOM (%s)".formatted(ctx));
+        final org.cyclonedx.model.Bom cdxBom = parseBom(ctx, bomFile);
+
+        // Keep track of which BOM ref points to which component identity.
+        // During component and service de-duplication, we'll potentially drop
+        // some BOM refs, which can break the dependency graph.
+        final var identitiesByBomRef = new HashMap<String, ComponentIdentity>();
```

Comment:
do we need this to be thread safe?

Reply:
Nope, the event system will create a new task instance for every event, so task instances are never accessed by multiple threads. 

The`Map` is also [local](https://www.baeldung.com/java-local-variables-thread-safe), and not a class-level variable. So even if multiple threads were sharing a task instance, each invocation of `processBom` would have its own `identitiesByBomRef`.